attribute-service-client/src/main/java/org/hypertrace/core/attribute/service/client/AttributeServiceClient.java

```diff
@@ -94,4 +96,12 @@ public void deleteSourceMetadata(
       Map<String, String> headers, AttributeSourceMetadataDeleteRequest req) {
     execute(headers, () -> blockingStub.deleteSourceMetadata(req));
   }
+
+  public void shutDown() throws InterruptedException {
+    Channel channel = blockingStub.getChannel();
```

Comment:
we have two public constructors including this one `AttributeServiceClient(Channel channel)`, so handling this way? Do you have any other suggestions?

Reply:
Mmh. Never had to look into this. I guess you are the pioneer.

Reply:
@aaron-steinfeld brought this to my attention https://github.com/hypertrace/hypertrace-core-graphql/blob/main/hypertrace-core-graphql-grpc-utils/src/main/java/org/hypertrace/core/graphql/utils/grpc/DefaultGrpcChannelRegistry.java

Reply:
So we do this in gql - https://github.com/hypertrace/hypertrace-core-graphql/blob/main/hypertrace-core-graphql-grpc-utils/src/main/java/org/hypertrace/core/graphql/utils/grpc/DefaultGrpcChannelRegistry.java

I'd recommend removing the interfaces for creating a client without a channel - that way the server is responsible for calling and managing its channels rather than each client. That channel param can also likely be switched to a ManagedChannel.

Reply:
The second part (That channel param can also likely be switched to a ManagedChannel) I got it, but in the first part, are you suggesting not to have `AttributeServiceClient(String host, int port)` and this one `AttributeServiceClient(AttributeServiceClientConfig attributeServiceClientConfig)`? 

And keeping `AttributeServiceClient(ManagedChannel channel)`, right?

Reply:
I also need a similar `shutdown` interface for entity service related clients too -https://github.com/hypertrace/entity-service/tree/main/entity-service-client/src/main/java/org/hypertrace/entity

Reply:
That would be my suggestion yes. Replacing the call is trivial enough (`AttributeServiceClient(ManagedChannelBuilder.forAddress...build())`, but it (intentionally) puts the burden on the caller to manage their channels, rather than the client. The client then can be agnostic to things like channel lifecycle and reuse which are service-specific.

Reply:
A little about grpc channel best practices here: https://groups.google.com/g/grpc-io/c/jc-cZak29V4

Reply:
Specifically, the discussion about cannel reuse:

> Reading another thread made me want to tweak my statement:
> Only in very heavy throughput (when we start talking about trying to fully saturate a server NIC) should you even begin to consider do something else.
> 
> Even in that case you probably shouldn't use multiple Channels. Instead, you would want a load balancer that might make multiple connections to the backend. Use different channels when you are contacting different backends. That is the only form of "channel pool" you might want: to look up a single shared Channel per backend. .

So I'm not saying we need to do that as part of this change at all (that's a service rather than client change anyway), but by taking a channel as the client builder argument, it encourages consuming services to better manage their channels rather than treating them as an implementation detail of the client.

Reply:
> I also need a similar `shutdown` interface for entity service related clients too -https://github.com/hypertrace/entity-service/tree/main/entity-service-client/src/main/java/org/hypertrace/entity

All clients do need this change, but the shutdown interface isn't needed if we pass in the channel - it would be shutdown from outside.

Reply:
Now, I got it (was confused with the word `server`). I will make changes, and update the PR.